src/full_node/full_node.py

```diff
@@ -1363,3 +1376,297 @@ async def respond_transaction(
+        """
+        # TODO(Florin): Use a cache for recent proofs, and ignore proof if it was already received.
+        if vdf_proof.witness_type > 0 or not vdf_proof.normalized_to_identity:
+            return False
+        if not vdf_proof.is_valid(self.constants, ClassgroupElement.get_default_element(), vdf_info):
+            return False
+        header_block = await self.blockchain.get_header_block_by_height(height)
+        if header_block is None:
+            return False
+        return await self._needs_compact_proof(vdf_info, header_block, field_vdf)
+
+    async def _replace_proof(
+        self,
+        vdf_info: VDFInfo,
+        vdf_proof: VDFProof,
+        height: uint32,
+        field_vdf: FieldVDF,
+    ):
+        full_blocks = await self.block_store.get_full_blocks_at([height])
+        assert len(full_blocks) > 0
+        block = full_blocks[0]
+        new_block = None
+        block_record = self.blockchain.height_to_block_record(height)
+
+        if field_vdf == FieldVDF.CC_EOS_VDF:
+            for index, sub_slot in enumerate(block.finished_sub_slots):
+                if sub_slot.challenge_chain.challenge_chain_end_of_slot_vdf == vdf_info:
+                    new_proofs = dataclasses.replace(sub_slot.proofs, challenge_chain_slot_proof=vdf_proof)
+                    new_subslot = dataclasses.replace(sub_slot, proofs=new_proofs)
+                    new_finished_subslots = block.finished_sub_slots
+                    new_finished_subslots[index] = new_subslot
+                    new_block = dataclasses.replace(block, finished_sub_slots=new_finished_subslots)
+                    break
+        if field_vdf == FieldVDF.ICC_EOS_VDF:
+            for index, sub_slot in enumerate(block.finished_sub_slots):
+                if (
+                    sub_slot.infused_challenge_chain is not None
+                    and sub_slot.infused_challenge_chain.infused_challenge_chain_end_of_slot_vdf == vdf_info
+                ):
+                    new_proofs = dataclasses.replace(sub_slot.proofs, infused_challenge_chain_slot_proof=vdf_proof)
+                    new_subslot = dataclasses.replace(sub_slot, proofs=new_proofs)
+                    new_finished_subslots = block.finished_sub_slots
+                    new_finished_subslots[index] = new_subslot
+                    new_block = dataclasses.replace(block, finished_sub_slots=new_finished_subslots)
+                    break
+        if field_vdf == FieldVDF.CC_SP_VDF:
+            assert block.challenge_chain_sp_proof is not None
+            new_block = dataclasses.replace(block, challenge_chain_sp_proof=vdf_proof)
+        if field_vdf == FieldVDF.CC_IP_VDF:
+            new_block = dataclasses.replace(block, challenge_chain_ip_proof=vdf_proof)
+        assert new_block is not None
+        await self.block_store.add_full_block(new_block, block_record)
```

Comment:
I think we should add a field in the DB that says: fully_compactified: True.
Then when doing the scan, we don't have to re-read 10,000,000 blocks from disk again. 
I think it's an absolute must.